polar-core/src/kb.rs

```diff
@@ -75,4 +85,102 @@ impl KnowledgeBase {
     pub fn is_constant(&self, name: &Symbol) -> bool {
         self.constants.contains_key(name)
     }
+
+    pub fn add_source(&mut self, source: Source) -> PolarResult<u64> {
+        let src_id = self.new_id();
+        if let Some(ref filename) = source.filename {
+            self.check_file(&source.src, &filename)?;
+            self.loaded_content
+                .insert(source.src.clone(), filename.to_string());
+            self.loaded_files.insert(filename.to_string(), src_id);
+        }
+        self.sources.add_source(source, src_id);
+        Ok(src_id)
+    }
+
+    pub fn clear_sources(&mut self) {
+        self.rules.clear();
+        self.sources = Sources::default();
+        self.inline_queries.clear();
+        self.loaded_content.clear();
+        self.loaded_files.clear();
+    }
+
+    pub fn remove_file(&mut self, filename: &str) -> Option<String> {
+        self.loaded_files
+            .get(filename)
+            .cloned()
+            .map(|src_id| self.remove_source(Some(filename.to_string()), src_id))
+    }
+
+    pub fn remove_source(&mut self, filename: Option<String>, source_id: u64) -> String {
```

Comment:
Why is filename optional?

Reply:
Does this function need to be public? Can we just call `remove_file` from `polar.rs` instead?

Reply:
Yes. And  we can merge these together.

Reply:
Okay, we still need this. The problem is if you load a raw string in that has errors, then you need the source in the KB so you can print nice errors, but then you want to remove the source. But there's no filename.

Reply:
Okay can we make this function just take `source_id` as an argument then so that it's clear when to use `remove_source` and when to use `remove_file`?

We can look at the values in `loaded_files` to clean up the filenames without needing the filename to be passed in.

Reply:
Yeah, that seems reasonable. Let me look.

Reply:
Good suggestion. Removed a redundant assert too.